docs: add READMEs for Slurm integration interface implementations (#239)
charmed-hpc/slurm-charms: internal/slurmd-interface/README.md

# `charmed-slurm-slurmd-interface`

## Usage

This package provides the integration interface implementation for the `slurmd` interface.
It enables `slurmd` (Slurm compute daemon) applications to exchange data with the `slurmctld` controller.
`slurmd` receives authentication secrets and controller addresses, and provides partition

> configuration back to `slurmctld`.

Reviewer: ```suggestion
This package provides the integration interface implementation for the `slurmd` interface. It enables `slurmd` (Slurm compute daemon) applications to exchange data with the `slurmctld` controller, such as controller addresses and partition configuration.
```

Reviewer: No reason to spell it out here and immediately below, and keeps the pattern and flow more consistent across the READMEs.

To install, add `charmed-slurm-slurmd-interface` to your Python dependencies.
Then in your Python code, import as:

```python
from charmed_slurm_slurmd_interface import (
    ComputeData,
    SlurmdProvider,
    SlurmdRequirer,
    controller_ready,
    partition_ready,
)
```

## Direction

The `slurmd` interface implements a provider/requirer pattern.
The Provider is the `slurmd` application that provides partition data to `slurmctld` and receives controller data.

> The Requirer is the `slurmctld` application that provides controller data and consumes partition configuration from `slurmd`.

Reviewer: ```suggestion
The `slurmd` interface implements a provider/requirer integration pattern. The provider-side application, `slurmd`, provides the integration endpoint that the requirer-side application, `slurmctld`, needs. `slurmctld` then supplies authentication secrets and controller data to `sackd` through the databag associated with the integration instance. `slurmd` then provides partition configuration data that `slurmctld` consumes. 
```

```mermaid
flowchart TD
    Provider -- partition --> Requirer

>     Requirer -- auth_secret_id, controllers --> Provider

Reviewer: ```suggestion
   A("slurmd\n(provider)") -- endpoint, partition config --> B("slurmctld\n(requirer)")
   B("slurmctld\n(requirer)") -- auth_secret_id, controllers --> A("slurmd\n(provider)") 
```

Author: Same as my other comments. We shouldn't have `endpoint` in the integration databag here since it's not exchanged through the application databag.

```

## Behavior

The `slurmctld` requirer provides controller data (authentication secret and controller addresses) to the `slurmd`
provider. In turn, the `slurmd` provider publishes its partition configuration so that `slurmctld` can incorporate

> compute nodes into the cluster configuration.

Reviewer: Can be rolled into Direction section above; I believe my suggestion covers most of it.

## Usage

This package provides the integration interface implementation for the `slurmd` interface.
It enables charmed applications providing the `slurmd` service (Slurm compute daemon)
to exchange partition and Slurm management data with charmed applications that require the

> `slurmd` service.

Reviewer: ```suggestion
It enables charmed applications that provide the `slurmd` service (Slurm compute daemon)
to exchange partition and Slurm management data with charmed applications that require the
`slurmd` service.
```

Reviewer: Just so 'provide' and 'require' are in the same word tense

The `slurmd` requirer provides controller data (authentication secret and controller addresses) to
the `slurmd` provider. In turn, the `slurmd` provider publishes its partition configuration so

> that the `slurmd` requirer can incorporate compute nodes into the cluster configuration.

Reviewer: ```suggestion
The `slurmd` requirer supplies controller data (authentication secret and controller addresses) to
the `slurmd` provider. In turn, the `slurmd` provider publishes its partition configuration so
that the `slurmd` requirer can incorporate compute nodes into the cluster configuration.
```
